frontend/src/redux/middleware.js

```diff
@@ -0,0 +1,55 @@
+import { addWidget } from '@metacell/geppetto-meta-client/common/layout/actions';
+import { WidgetStatus } from '@metacell/geppetto-meta-client/common/layout/model';
+import { v4 as uuidv4 } from 'uuid';
+import { ADD_VIEWER } from './actions/viewers';
+import { ADD_DEVSTAGE } from './actions/devStages';
+import { raiseError, waitData } from './actions/misc';
+import { DevStageService } from '../services/DevStageService';
+
+const devStagesService = new DevStageService();
+
+const middleware = (store) => (next) => (action) => {
+  function widgetFromViewerSpec(viewerSpec) {
```

Comment:
would move this function outside of the middleware scope, makes the middleware more readable imho